### PR TITLE
Set Up Travis CI

### PR DESCRIPTION
**Pull request by abolinsky**
Since Travis does not support Ubuntu 16.04, we are running the build, tests, and scan within a docker container running 16.04. The .travis.yml file is smart enough to pull in the other repo as a dependency, first checking if there is a branch of the same name, then defaulting to **rr-devel** in the event there isn't. 

Builds will pass on the first run if only one repo has been changed. If both have been changed, and you committed to both branches around the same time, then the builds should pass. If you make changes to both and commit the changes to both branches with sufficient delay between the commits, then when the first build attempts to merge in dependencies, it will not find them and the build will fail.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,76 +1,100 @@
...
+        python-catkin-pkg 
+        python-rosdep 
+        python-wstool 
+        python-catkin-tools 
+        ros-kinetic-catkin 
+        ros-kinetic-rostest 
+        wget 
+        curl && 
+    source /opt/ros/kinetic/setup.bash && 
+    catkin_init_workspace &&
```
Comment (gauthamm) [inline .travis.yml]: we are using catkin tools, so this should be `catkin init`

File: .travis.yml [hunk 2]
```diff
@@ -1,76 +1,100 @@
...
+        ros-kinetic-catkin 
+        ros-kinetic-rostest 
+        wget 
+        curl && 
+    source /opt/ros/kinetic/setup.bash && 
+    catkin_init_workspace && 
+    cd ~/catkin_ws && 
+    catkin build && 
+    source devel/setup.bash && 
+    cd ~/catkin_ws/src &&
```
Comment (gauthamm) [inline .travis.yml:44]: We could make a public image available on a general dockerhub account that contains all the stuff here. In install we just need to do a `rosdep update && apt-get update`. I do think having them in one file here is cleaner though.

File: .travis.yml [hunk 3]
```diff
@@ -1,76 +1,100 @@
...
-  - sudo sh -c "echo \"deb http://packages.ros.org/ros/ubuntu $ROS_CI_DESKTOP main\" > /etc/apt/sources.list.d/ros-latest.list"
-  - sudo apt-key adv --keyserver hkp://ha.pool.sks-keyservers.net:80 --recv-key 421C365BD9FF1F717815A3895523BAEEB01FA116
-  - sudo apt-get update -qq
-  - sudo apt-get install -y python-catkin-pkg python-rosdep python-wstool ros-$ROS_DISTRO-catkin python-catkin-tools
-  - source /opt/ros/$ROS_DISTRO/setup.bash
-  # Prepare rosdep to install dependencies.
-  - sudo rosdep init
-  - rosdep update
+  - docker run -e BRANCH -e PYTHONPATH -d --name alica-test -v $TRAVIS_BUILD_DIR:/travis ros:kinetic-ros-core-xenial tail -f /dev/null
+  - docker ps
```
Comment (gauthamm) [inline .travis.yml]: Why is this needed?

File: .travis.yml [hunk 4]
```diff
@@ -1,76 +1,100 @@
...
-  - sudo rm /var/lib/dpkg/lock
-  - sudo sh -c "echo \"deb http://packages.ros.org/ros/ubuntu $ROS_CI_DESKTOP main\" > /etc/apt/sources.list.d/ros-latest.list"
-  - sudo apt-key adv --keyserver hkp://ha.pool.sks-keyservers.net:80 --recv-key 421C365BD9FF1F717815A3895523BAEEB01FA116
-  - sudo apt-get update -qq
-  - sudo apt-get install -y python-catkin-pkg python-rosdep python-wstool ros-$ROS_DISTRO-catkin python-catkin-tools
-  - source /opt/ros/$ROS_DISTRO/setup.bash
-  # Prepare rosdep to install dependencies.
-  - sudo rosdep init
-  - rosdep update
+  - docker run -e BRANCH -e PYTHONPATH -d --name alica-test -v $TRAVIS_BUILD_DIR:/travis ros:kinetic-ros-core-xenial tail -f /dev/null
```
Comment (gauthamm) [inline .travis.yml]: why are you tailing /dev/null?

File: .travis.yml [hunk 5]
```diff
@@ -1,76 +1,100 @@
...
-  - catkin run_tests && catkin_test_results
+  - docker exec -t alica-test bash -c "
+    source /opt/ros/kinetic/setup.bash && 
+    cd ~/catkin_ws && 
+    catkin build --no-status && 
+    source devel/setup.bash && 
+    catkin run_tests --no-status && catkin_test_results && 
+    if [[ \"$BRANCH\" = \"coverity_scan\" ]] ; then 
+        wget -qO- https://scan.coverity.com/download/cxx/linux64 
+            --post-data \"token=o6qYj0vj-t-K1PKQj5ugyg&project=rapyuta-robotics/alica\" |
```
Comment (gauthamm) [inline .travis.yml]: why alica in alica-supplementary?

Comment (gauthamm) [reply .travis.yml]: please make sure this token is valid on the general rapyuta account

File: .travis.yml [hunk 6]
```diff
@@ -1,76 +1,100 @@
...
+        wget -qO- https://scan.coverity.com/download/cxx/linux64 
+            --post-data \"token=o6qYj0vj-t-K1PKQj5ugyg&project=rapyuta-robotics/alica\" | 
+        tar xvz && 
+        python /travis/generate_coverity-analysis_script.py . && 
+        chmod +x coverity-analysis.sh && 
+        ./cov-analysis*/bin/cov-build --dir cov-int ./coverity-analysis.sh && 
+        tar czvf cs.tgz cov-int && 
+        curl 
+            --form token=o6qYj0vj-t-K1PKQj5ugyg 
+            --form email=alexander.bolinsky@rapyuta-robotics.com
```
Comment (gauthamm) [inline .travis.yml]: general rapyuta account, it might make sense to make this an environment variable.

File: .travis.yml [hunk 7]
```diff
@@ -1,76 +1,100 @@
...
+    if [[ \"$BRANCH\" = \"coverity_scan\" ]] ; then 
+        wget -qO- https://scan.coverity.com/download/cxx/linux64 
+            --post-data \"token=o6qYj0vj-t-K1PKQj5ugyg&project=rapyuta-robotics/alica\" | 
+        tar xvz && 
+        python /travis/generate_coverity-analysis_script.py . && 
+        chmod +x coverity-analysis.sh && 
+        ./cov-analysis*/bin/cov-build --dir cov-int ./coverity-analysis.sh && 
+        tar czvf cs.tgz cov-int && 
+        curl 
+            --form token=o6qYj0vj-t-K1PKQj5ugyg
```
Comment (gauthamm) [inline .travis.yml]: might make sense to make it an environment variable

File: .travis.yml [hunk 8]
```diff
@@ -1,76 +1,100 @@
...
+        chmod +x coverity-analysis.sh && 
+        ./cov-analysis*/bin/cov-build --dir cov-int ./coverity-analysis.sh && 
+        tar czvf cs.tgz cov-int && 
+        curl 
+            --form token=o6qYj0vj-t-K1PKQj5ugyg 
+            --form email=alexander.bolinsky@rapyuta-robotics.com 
+            --form file=@cs.tgz 
+            --form version=\"1.0\" 
+            --form description=\"ALICA\" 
+            https://scan.coverity.com/builds?project=rapyuta-robotics%2Falica ;
```
Comment (gauthamm) [inline .travis.yml]: alica-supplementary

**Review by hendrik-skubch — APPROVED** [review]
Looks all good to me.

File: .travis.yml [hunk 9]
```diff
@@ -1,76 +1,100 @@
...
+    source /opt/ros/kinetic/setup.bash && 
+    cd ~/catkin_ws && 
+    catkin build --no-status && 
+    source devel/setup.bash && 
+    catkin run_tests --no-status && catkin_test_results && 
+    if [[ \"$BRANCH\" = \"coverity_scan\" ]] ; then 
+        wget -qO- https://scan.coverity.com/download/cxx/linux64 
+            --post-data \"token=o6qYj0vj-t-K1PKQj5ugyg&project=rapyuta-robotics/alica\" | 
+        tar xvz && 
+        python /travis/generate_coverity-analysis_script.py . &&
```
Comment (gauthamm) [inline .travis.yml]: might be better to explicitly pass the catkin_ws directory here.

File: .travis.yml [hunk 10]
```diff
@@ -1,76 +1,100 @@
...
+    if [[ -f /travis/dependencies.rosinstall ]] ; then 
+        sed -i -e \"s/rr-devel/$BRANCH/g\" /travis/dependencies.rosinstall; 
+        if wstool merge -y /travis/dependencies.rosinstall ; then 
+            sed -i -e \"s/$BRANCH/rr-devel/g\" /travis/dependencies.rosinstall; 
+            wstool merge -y /travis/dependencies.rosinstall; 
+        fi; 
+    fi && 
+    wstool up && 
+    cd ~/catkin_ws && 
+    rosdep install -y --from-paths src --ignore-src --rosdistro kinetic"
```
Comment (gauthamm) [inline .travis.yml]: replace kinetic with $ROS_DISTRO

File: .travis.yml [hunk 11]
```diff
@@ -1,76 +1,100 @@
...
+    rosdep update && 
+    mkdir -p ~/catkin_ws/src && 
+    cd ~/catkin_ws/src && 
+    apt-get update && 
+    apt-get install -y 
+        python-catkin-pkg 
+        python-rosdep 
+        python-wstool 
+        python-catkin-tools 
+        ros-kinetic-catkin
```
Comment (gauthamm) [inline .travis.yml]: Might be better to change all kinetic references to $ROS_DISTRO

File: .travis.yml (hunk 6 above)
Comment (gauthamm) [reply .travis.yml]: If this is going to be merged upstream it makes sense to comeup with a general email account used by the Alica team.

File: .travis.yml (hunk 1 above)
Comment (gauthamm) [reply .travis.yml]: On second thought I am not sure if catkin tools works properly with travis for evaluating tests.

File: .travis.yml (hunk 3 above)
Comment (abolinsky) [reply .travis.yml]: It's not necessary. Just for debugging. I'll get rid of it.

File: .travis.yml (hunk 4 above)
Comment (abolinsky) [reply .travis.yml]: If docker run is not provided a command then the container shuts down. tailing /dev/null allows the container to stay open so I can pass it commands later as I'm doing in the several calls to exec.

File: .travis.yml (hunk 5 above)
Comment (abolinsky) [reply .travis.yml]: Believe it or not, alica depends on supplementary, and supplementary depends on alica. It's dumb, and eventually we will want to completely decouple them.

Furthermore, the token will be valid even after changing the email associated with the Coverity account.

File: .travis.yml (hunk 6 above)
Comment (abolinsky) [reply .travis.yml]: Made them environment variables!

File: .travis.yml (hunk 7 above)
Comment (abolinsky) [reply .travis.yml]: done

File: .travis.yml (hunk 9 above)
Comment (abolinsky) [reply .travis.yml]: can you explain?
I am using the TRAVIS_BUILD_DIR environment variable to mount the repository that is being built into docker as /travis.

File: .travis.yml (hunk 11 above)
Comment (abolinsky) [reply .travis.yml]: done

File: .travis.yml (hunk 8 above)
Comment (abolinsky) [reply .travis.yml]: I've taken coverity scan out of the supplementary travis.yml to avoid confusion.

File: .travis.yml (hunk 9 above)
Comment (gauthamm) [reply .travis.yml]: instead of using `.` to mention current directory, it would be better to pass the actual path to catkin_ws

**Review by gauthamm — APPROVED** [review]
